---
title: "Step 3: ATAC-Seq-Data-Analysis-QC on all cells"
subtitle: "Step 3.2.1: Visualization of ATAC-seq-QC on all cells using ln(1+x) normalization"
date: "2025-02-11"
author: Xenia Sirodzha
output:
  html_document: 
    toc: yes
    toc_depth: 4
    toc_float: yes
    df_print: paged
---

As the workflow for ATAC-seq QC was established in part 1, I performed this workflow on the entire data set. However, performing pseudo-bulking and normalization and subsequent visualization with the data from all cells is very computationally intensive. Therefore, visualization was performed separately from the pseudo-bulking and manual ln(1+x) normalization here. 

# Libraries

These are the packages required for the code below.
```{r libraries}
library(tidyverse) 
library(binom)
library(lubridate)
library(broom)
library(scales) #for scale transformations
library(cowplot) #arranging plots 
library(ggpubr)
library(ggridges)
library(ggthemes) #stylising plots
library(ggplot2)
library(ggsci) #color palettes
library(readxl) #for loading excel files
library(Seurat) #for rna-seq data, atac-seq data
library(Signac) #for atac seq data
```

# 1 Data loading

For visualization, the manually normalized data set from "02-11-ATAC-Seq-Analysis-Step-3.1.1-log 1+x on all celltypes" was loaded into the session.
```{r data loading}
readRDS("Data/psd.bulk.zfish_atac.lg1x.rds") -> psd.bulk.zfish_atac.lg1x
```


# 2 Data wrangling for visualization

After normalization, I prepared the data to plot with tidyverse. First, I converted the matrix to a data frame, then I calculated the values which were plotted as well (sum of pk_counts, coefficient of variance (cv)). 

```{r data wrangling for visualization}
#procedure
as.data.frame(psd.bulk.zfish_atac.lg1x)%>% 
  pivot_longer(cols=!contains("rownames"), ,values_to = "pk_counts", names_to = "celltype") %>% 
  mutate(pk_count_sum= sum(pk_counts)) %>% 
  mutate(cv=sd(pk_counts) / mean(pk_counts)) %>% 
  group_by(celltype) %>% 
  mutate(pk_count_sum_celltype= sum(pk_counts)) %>% 
  mutate(cv_celltype=sd(pk_counts) / mean(pk_counts)) %>% 
  mutate(all="all celltypes")-> df_psd.bulk.zfish_atac.lg1x

#output
df_psd.bulk.zfish_atac.lg1x 
```

# 3 Visualization

Subsequently, the wrangled data was visualized analogous to the script in step 2 "02-06-ATAC-Seq-Analysis-Step-2.1 seurat norm". 

## 3.1 Histograms after pseudo-bulking by celltype

Then, the distribution of peaks overall and by celltype was visualized below.
```{r histograms peak count distribution}

# all clusters
df_psd.bulk.zfish_atac.lg1x %>% 
 ggplot(aes(x=pk_counts))+
 geom_histogram(binwidth = 0.2, color="gray15", alpha=0.5)+
 scale_y_continuous(breaks = scales::breaks_pretty(n=10))+
  coord_equal(ratio=2)+
  coord_fixed(ratio = 0.000075)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") ->dis.counts.1

#by celltype
 ggplot(df_psd.bulk.zfish_atac.lg1x, aes(x=pk_counts, fill = celltype))+
  geom_histogram(binwidth = 0.2, color="gray15", alpha=0.8)+
  scale_y_continuous(breaks = scales::breaks_pretty(n=5))+
  facet_wrap(facets = "celltype")+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") ->dis.counts.cell.1

#side by side 
dis.counts.1+dis.counts.cell.1
```

Apart from the digestive system cells and the erythroid lineage cells, all peaks in all celltypes are normally distributed. 


## 3.2 Sum of all peaks

The sum of all peaks is 22,403,833.
```{r extracting value sum of all peaks}
unique(df_psd.bulk.zfish_atac.lg1x$pk_count_sum)
```
Then, the sum of all peaks overall and by celltype was visualized below.
```{r bar plot sum of all peaks}
# all clusters
df_psd.bulk.zfish_atac.lg1x %>% 
  ungroup() %>% 
  distinct(all, pk_count_sum) %>% 
  ggplot(aes(x=all, y=pk_count_sum))+
  geom_col(alpha=0.8, color="gray10")+
  scale_y_continuous(n.breaks=10)+
  coord_fixed(ratio = 0.0000005)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> sum.peaks.all.1

# by cell type
df_psd.bulk.zfish_atac.lg1x %>% 
  group_by(celltype) %>% 
  distinct(pk_count_sum_celltype, pk_count_sum) %>% 
  ggplot(aes(x=celltype, y=pk_count_sum_celltype, fill=celltype))+
  geom_col(alpha=0.8, color="gray10")+
   scale_y_continuous(n.breaks=20)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        axis.text.x = element_text(angle = 45, hjust = 1),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> sum.peaks.celltype.1

# side by side
sum.peaks.all.1|sum.peaks.celltype.1
```

Most peaks range 70,000 to 125,000 per cell type. Epiblast cells have the lowest sum of all peaks (~40,000), indicating that in this data set, they are either a cell type with little cells, low gene expresssion or both.

## 3.4 Coefficient of variation 

Next, the coefficient of variation was visualised in a bar plot. 
```{r bar plot coefficient of variation}
df_psd.bulk.zfish_atac.lg1x %>% 
  ungroup() %>% 
  distinct(all, cv) %>% 
  ggplot(aes(x=all, y=cv ))+
  geom_col(alpha=0.8, color="gray10")+
  scale_y_continuous(n.breaks=5)+
  coord_fixed(ratio = 10)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> cv.all.1

df_psd.bulk.zfish_atac.lg1x %>% 
  group_by(celltype) %>% 
  distinct(cv_celltype, cv) %>% 
  ggplot(aes(x=celltype, y=cv_celltype, fill=celltype))+
  geom_col(alpha=0.8, color="gray10")+
  scale_y_continuous(n.breaks=10)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        axis.text.x = element_text(angle = 45, hjust = 1),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") ->cv.celltype.1

cv.all.1|cv.celltype.1
```

Most cell types possess a coefficient of variation similar to the overall coefficient variation of 0.44. Their coefficients of variation range from 0.23 to 0.42. 

## 3.5 Count distribution visualized in a violin plot by cell type

Then, the count distribution was visualised with a violin plot as well as a violin plot with a boxplot drawn on top
```{r violin plot counts}
df_psd.bulk.zfish_atac.lg1x %>% 
  mutate(all="all") %>% 
 ggplot( aes( x=all, y=pk_counts))+
  geom_violin(color="grey20", alpha=0.3, fill="gray30", trim=F)+
  scale_y_continuous(n.breaks=10)+
  coord_fixed(ratio = 0.5)+
  theme_bw()+ theme(text = element_text(size = 4, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> v.plot.all.1


 ggplot(df_psd.bulk.zfish_atac.lg1x, aes(x=celltype, y=pk_counts, fill = celltype))+
  geom_violin(color="grey20", alpha=0.3, trim=F)+
   coord_equal(ratio = 0.8)+
   scale_y_continuous(n.breaks=10)+
  theme_bw()+ theme(text = element_text(size = 4, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> v.plot.cells.1

# side by side
v.plot.all.1|v.plot.cells.1
```


The data is hardly visible so I made the following adjustments: I added a boxplot and scale limits set to 0 and 0.1

The resulting plot can be seen down below. 
```{r violin plot counts plus boxplot}
df_psd.bulk.zfish_atac.lg1x %>% 
  mutate(all="all") %>% 
 ggplot( aes( x=all, y=pk_counts))+
  geom_violin(color="grey20", alpha=0.3, fill="gray30", trim=F)+
  geom_boxplot(width = 0.07, alpha=0.3)+
  stat_boxplot(geom ='errorbar', width=0.2, position=position_dodge(0.5))+
  scale_y_continuous(n.breaks=10)+
  coord_fixed(ratio = 0.8)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> v.plot.all.box.1


 ggplot(df_psd.bulk.zfish_atac.lg1x, aes(x=celltype, y=pk_counts, fill = celltype))+
  geom_violin(color="grey20", alpha=0.3, trim=F)+
  geom_boxplot(width = 0.07, alpha=0.3)+
  stat_boxplot(geom ='errorbar', width=0.2, position=position_dodge(0.5))+
   scale_y_continuous(n.breaks=10)+
  theme_bw()+ theme(text = element_text(size = 9, family="sans", color="black"), #stylisation
        axis.text = element_text(size=8, color="black"),
        axis.text.x = element_text(angle = 65, hjust = 1),
        axis.title=element_text(size=9, face = "bold"),
        plot.subtitle=element_text(size=10.5),
        plot.title=element_text(size=11, face = "bold"), 
        legend.position ="none") -> v.plot.cells.box.1
 
 v.plot.all.box.1|v.plot.cells.box.1
```

Overall, most peak counts range from 0.5 to 5.6 and have a median of 2.6. This shows that this sample was normalized well.

To view the violin plots per cell type in more detail, the plots were displayed separately as well. 

```{r violin plot counts plus boxplot focus by celltype}
v.plot.cells.box.1
```

Most peak counts are normally distributed by cell type. However, digestive system cells and erythroid lineage cells are not normally distributed. Blastomere and epiblast cells have the lowest peak count median (~1.0). Of the normally distributed cell types YSL and anterior posterior cells have the highest median (~3.0). 

To summarize, the values were normalized well by manual ln(1+x) transformation. In addition, digestive system and erythroid lineage cells are not normally distributed. Therefore, they are not as suitable for modelling as the other cell types. 

# Session info

Information about the coding environment can be found below.
```{r session info}
sessionInfo()
```
