---
title: "Step 3 ATAC-Seq-Data-Analysis-QC on all cells"
subtitle: "Step 3.1.1: Pseudo-Bulking and normalization using ln(1+x) normalization"
date: "2025-02-11"
author: Xenia Sirodzha
output:
  html_document: 
    toc: yes
    toc_depth: 4
    toc_float: yes
    df_print: paged
---

As the workflow for ATAC-seq QC was established in step 2, I performed this workflow on the entire data set. However, performing pseudo-bulking and normalization and subsequent visualization with the data from all cells is very computationally intensive. Therefore, pseudo-bulking and manual ln(1+x) normalization was performed separately from the visualization here. 

# Libraries

These are the packages required for the code below.
```{r libraries}
library(tidyverse) 
library(binom)
library(broom)
library(Seurat) #for rna-seq data, atac-seq data
library(Signac) #for atac seq data
```


# 1. Data loading and inspecting

First, I loaded these data sets into R to work with them. The data sets were graciously extracted and compiled by Lauren Saunders from [Lin et al. 2023](https://www.nature.com/articles/s41597-023-02373-y). 

```{r data loading}
readRDS("Data/Lin_2023_zfish_snATAC-seq/Cell-type_Peak_Matrix.rds") -> zfish_snATAC_seq_pk_mtrx #count matrix
read_csv("Data/Lin_2023_zfish_snATAC-seq/atac_all.metaData.csv")-> zfish_mta_data # meta data
```

Then, I inspected the data sets with `class()` and `dim()`. 
The matrix `zfish_snATAC_seq_pk_mtrx` contains the peaks as rownames and columns as cell names and counts as entries. As this is a sparse matrix, the 0s are written as dots. In addition, the data frame `zfish_mta_data` contains the meta data.

```{r inspecting matrix and meta data}
unique(zfish_mta_data$Clusters) #how many clusters in total
class(zfish_snATAC_seq_pk_mtrx)
class(zfish_mta_data)
dim(zfish_snATAC_seq_pk_mtrx)
zfish_snATAC_seq_pk_mtrx[1:10, 1:2]
zfish_mta_data
```

For further analysis in this ATAC-Seq-QC to determine whether this data set can be used to create a model which predicts chromatin accessibility based on genomic sequence alone, I will create a seurat object with the code below.

```{r creating seurat object}
zfish_atac=CreateSeuratObject(counts = zfish_snATAC_seq_pk_mtrx, assay = "atac", meta.data = zfish_mta_data) #combines matrix and meta data into a seurat object
```

Here is an overview of the seurat object, obtained with the `glimpse` function.

```{r inspecting seurat object}
glimpse(zfish_atac)
```

# 2. Data transformation

Next the data was transformed.
Overall, it was pseudo-bulked,log-normalized manually and saved as an RDS file for later use. To generate the matrix `psd.bulk.zfish_atac`, which contains all un-normalized counts per cell type, I pseudo-bulked the count data from the `zfish_atac` object by setting the arguments to: `return.seurat=F` and no specified normalization method, which is the default of this function. 

```{r pseudobulking}

#pseudobulking

psd.bulk.zfish_atac= AggregateExpression(zfish_atac, group.by = "celltype") 

#output

glimpse(psd.bulk.zfish_atac)
```


After that, this matrix was normalized manually, by dividing the sum of the counts by the number of cells in each celltype then, by multiplying that number by a 1000 for the average in 1000 cells and lastly by transforming the matrix with ln(x+1).

To achieve this, the number of cells of each cell type had to be extracted first.
```{r extracting number of cells}
zfish_atac@meta.data %>% 
  group_by(celltype) %>% 
  tally() %>% 
  pivot_wider(names_from = celltype, values_from = n) %>% 
  as.matrix() -> n_cells
n_cells
```

Then, I employed the following loop to extract the correct number of cells of each celltype to divide the counts of the matrix containing the pseudo-bulk with. 
```{r dividing number of cells}
##Dividing the counts by the number of cells in each cell type

psd.bulk.zfish_atac$atac[1:5,1:5] #input

for (i in seq_along(colnames(n_cells))) {
  ct <- colnames(n_cells)[i]  #extracting cell type names
  
#loop along cell type names in the pseudo bulk matrix
  for (j in seq_along(colnames(psd.bulk.zfish_atac$atac))) {
    atc <- colnames(psd.bulk.zfish_atac$atac)[j]  
    
#if the colnames match divide by the number of cells
    if (atc == ct) {
      psd.bulk.zfish_atac$atac[,j] <- (psd.bulk.zfish_atac$atac[, j] / n_cells[, i])*1000
      break  # stop if everything is matched 
    }
  }
}

psd.bulk.zfish_atac$atac[1:5,1:5] #output

```

After that, the matrix was ln(1+x) transformed in the manner seen below

```{r ln(1+x) transformation}
psd.bulk.zfish_atac=log(psd.bulk.zfish_atac$atac+1)

psd.bulk.zfish_atac[1:10,1:5]
```


Lastly, the transformed matrix was saved as a RDS file for further analysis.

```{r saving RDS, eval=FALSE}
#save as file

saveRDS(psd.bulk.zfish_atac, file = "Data/psd.bulk.zfish_atac.lg1x.rds")
```


# Session Info

Information about the coding environment can be found below.
```{r session info}
sessionInfo()
```